src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -4113,19 +4120,33 @@ bool MergeTreeData::getQueryProcessingStageWithAggregateProjection(
         return match;
     };
 
-    for (const auto & projection : metadata_snapshot->projections)
+    DataTypePtr partition_value_type = getPartitionValueType();
+
+    auto add_projection_candidate = [&](const ProjectionDescription & projection)
     {
         ProjectionCandidate candidate{};
         candidate.desc = &projection;
 
+        auto sample_block = projection.sample_block;
+        sample_block.insertUnique({ColumnString::create(), std::make_shared<DataTypeString>(), "_part"});
+        sample_block.insertUnique({ColumnString::create(), std::make_shared<DataTypeString>(), "_partition_id"});
+        sample_block.insertUnique({ColumnUUID::create(), std::make_shared<DataTypeUUID>(), "_part_uuid"});
+        sample_block.insertUnique({partition_value_type->createColumn(), partition_value_type, "_partition_value"});
+
+        auto sample_block_for_keys = projection.sample_block_for_keys;
+        sample_block_for_keys.insertUnique({ColumnString::create(), std::make_shared<DataTypeString>(), "_part"});
+        sample_block_for_keys.insertUnique({ColumnString::create(), std::make_shared<DataTypeString>(), "_partition_id"});
+        sample_block_for_keys.insertUnique({ColumnUUID::create(), std::make_shared<DataTypeUUID>(), "_part_uuid"});
+        sample_block_for_keys.insertUnique({partition_value_type->createColumn(), partition_value_type, "_partition_value"});
```

Comment:
Why do we need to add this virtual columns? Could you also add a comment here?

I am a little bit worried that we have several places where those virtual columns are mentioned. Now, if I add virtual column, should I also add it here? (And how would I know?)